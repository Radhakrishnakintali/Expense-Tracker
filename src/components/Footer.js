import React from "react";

export const Footer = () => {
    return (
        <div>
            <h5>
                Expense Tracker
            </h5>
        </div>
    )
}