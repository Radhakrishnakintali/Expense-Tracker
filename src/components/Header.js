import React from "react";
import { useTranslation }  from "react-i18next";

export const Header = () => {
    const { t } = useTranslation();
    //const { t } = jest ? { t: s => s } : useTranslation.useTranslation();
    return (
        <div style = {{width:"100%" , textAlign:"center", backgroundColor:"lightskyblue"}}>
            <h2 className="font-weight-normal mn-3">
                {t("expense_tracker")}
            </h2>
        </div>
    )
}