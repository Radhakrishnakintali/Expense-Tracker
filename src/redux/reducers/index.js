import { combineReducers } from "redux";
import account from './account';

const rootReducer = combineReducers({
    accounts: account
});

export default rootReducer;