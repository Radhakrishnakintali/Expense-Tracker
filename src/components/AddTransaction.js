import React, {useEffect, useState} from 'react';
import { bindActionCreators } from 'redux';
import { useDispatch } from 'react-redux';
import * as actions from '../redux/actions/accounts';
import useTranslation from "react-i18next";
import { popper } from '@popperjs/core';

export const AddTransaction = (props) => {
    // const { t } = useTranslation();
    const { t } = jest ? { t: s => s } : useTranslation.useTranslation();
    const [text, setText] = useState("");
    const [amount, setAmount] = useState(0);
    const [id, setId] = useState(0);
    const dispatch = useDispatch();
    const {addTransaction, updateTransaction} = bindActionCreators(actions, dispatch);

    useEffect(() => {
        if (props.transaction.length > 0) {
            console.log("running1: " +props.transaction);
            setId(props.transaction[0].id)
            setText(props.transaction[0].text);
            setAmount(props.transaction[0].amount);
        }
      }, [props]);

    const onSubmit = e => {
        e.preventDefault();
        if (id === 0) {
            const newTransaction = {
                id: Math.floor(Math.random() * 100000000),
                text,
                amount
            }
            addTransaction(newTransaction);
        } else {
            const oldTransaction = {
                id: id,
                text,
                amount
            }
            updateTransaction(oldTransaction);
        }

        setId(0);
        setText("");
        setAmount(0);
        props.clearData();
    }

    return (
        <div data-testid='addTransaction' style={{width:"40%"}}>
        <h3>{t("add_new_transaction")}</h3>
        <form  id="form" onSubmit={onSubmit}>
            <div className="form-control">
                <label htmlFor="text">{t("text")}</label>
                <input type="text" id="text" value = {text} onChange={(e) => setText(e.target.value)} placeholder={`${t("enter_text")}...`} />
            </div>
            <div className="form-control">
                <label htmlFor="amount"
                    >{t("amount")} <br />
                    ({t("negative")} - {t("expense")}, {t("positive")} - {t("income")})
                </label>
                <input type="number" id="amount" value = {amount} onChange={(e) => setAmount(e.target.value)} placeholder={`${t("enter_amount")}...`} />
            </div>
            <button data-testid='btnAddTransaction' style={{backgroundColor:"lightblue"}} className="btn">{t("add_transaction")}</button>
        </form>
      </div>
    );
}