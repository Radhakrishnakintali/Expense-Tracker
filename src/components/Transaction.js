import React, {useEffect} from "react";

import { useDispatch } from 'react-redux';
import * as actions from '../redux/actions/accounts';
import { bindActionCreators } from 'redux';

export const Transaction = (props) => {
    const sign = props.transaction.amount < 0 ? '-' : '+';
    const dispatch = useDispatch();
    const { deleteTransaction } = bindActionCreators(actions, dispatch);
    var updateTransaction = "";
    const getTransaction =(id) => {
        updateTransaction = props.transactions.filter(transaction => transaction.id === id)
        //localStorage.setItem('updateTransaction', transaction);
        //console.log(transaction);
        props.getData(updateTransaction);
    }
    useEffect(() => {
        localStorage.setItem('updateTransaction', JSON.stringify(updateTransaction))
      }, [updateTransaction])
    return (
        <li data-testid= {props.id} className= {props.transaction.amount < 0 ? "minus" : "plus"}>
            {props.transaction.text} <span>{sign}${Math.abs(props.transaction.amount)}</span><button data-testid={`btn${props.id}`} onClick={() => deleteTransaction(props.transaction.id)} className="delete-btn">x</button><button onClick={() => getTransaction(props.transaction.id)} type="button">
          Edit</button>
        </li> 
    )
}