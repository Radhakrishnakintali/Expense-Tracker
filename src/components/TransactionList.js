import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import  useTranslation  from "react-i18next";
import { Transaction } from './Transaction';
import { MysteriousSankey } from './MysteriousSankey';
import { AddTransaction } from './AddTransaction';


export const TransactionList = (props) => {
    const {accounts} = useSelector((accounts) => accounts);
    const [transactionData, setTransactionData] = useState({});
    //const { t } = useTranslation();
    const { t } = jest ? { t: s => s } : useTranslation.useTranslation();
     
    let count = 0;

    let nodes = accounts.transactions
            .map( transaction => 
                {
                    count ++;
                    return { "name": transaction.text };
                }
            )

    let links = accounts.transactions
            .map ((transaction, index) =>{
                count = count + 1
                if (transaction.amount < 0) 
                    return {
                        "source": 0,
                        "target": index + 1,
                        "value": Math.abs(transaction.amount)
                    }
                    return {
                        "source": index + 1,
                        "target": 0,
                        "value": transaction.amount
                    }
                }
            )
    nodes.splice(0, 0, {"name": "Budget"});

    const data = {
        "nodes": nodes,
        "links": links
    }

    const getData = (updateTransaction) => {
        setTransactionData(updateTransaction);
    }

    const clearData = () => {
        setTransactionData({});
    }

    return (
        <div>
            <div data-testid='transactionListData'>
                <div width="100%" >
                    <div className="font-weight-normal mb-2" style={{paddingTop:"20px"}}>{t("financial_flow_chart")}</div>
                    <svg style={{margin:"auto", display:"block"}} width="600px" height="600px">
                        {data && <MysteriousSankey data={data} width="600" height="600" />}
                    </svg>
                </div>
                <h3>{t("history")}</h3>
                <ul id="list" className="list">
                    {accounts.transactions.map((transaction,index) => (
                        <Transaction id={`transaction${index}`} key={transaction.id} transaction = {transaction} transactions={accounts.transactions}  getData={getData}/>
                    ))}
                </ul>
            </div>
            <AddTransaction transaction={transactionData} clearData={clearData}/>
        </div>
    )
}