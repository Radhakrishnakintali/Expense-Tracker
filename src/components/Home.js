import React from 'react';
//import { AddTransaction } from './AddTransaction';
import { TransactionList } from './TransactionList';

const Home = () => {
    return (
        <div style={{
            width: "100%"
        }}>
            <TransactionList />
            {/* <AddTransaction transaction={transaction}/> */}
        </div>
    )
};

export default Home;