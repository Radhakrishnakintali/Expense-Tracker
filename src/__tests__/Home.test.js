import {render, fireEvent} from '@testing-library/react';
import Home from '../components/Home';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore } from 'redux';
import rootReducer from '../redux/reducers/index';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { configure } from "enzyme";
import { AddTransaction } from '../components/AddTransaction';
import Adapter from 'enzyme-adapter-react-15';

configure({adapter: new Adapter()});
const middlewares = [thunk];

const mockStore = configureMockStore(middlewares);
const store = mockStore();



const initialState = {
    transactions: [{ id: 1, text: 'Flower', amount: -20 },
      { id: 2, text: 'Salary', amount: 300 },
      { id: 3, text: 'Book', amount: -10 },
      { id: 4, text: 'Camera', amount: 150 }]
}


function renderwithRedux(ui, {initialState, store = createStore(rootReducer,initialState,applyMiddleware(thunk))} = {}) {
    return {
    ...render(<Provider store={store}>{ui}</Provider>), store
    }
}

beforeEach(() => { // Runs before each test in the suite
    store.clearActions();
});

test('financial_flow_chart', () => {
    const { getByText } = renderwithRedux(<Home/>);
    const homeElement = getByText('financial_flow_chart');
    //console.log(homeElement);
    expect(homeElement).toBeInTheDocument();
})

test('should render Transaction List', () => {
    const {getByTestId} = renderwithRedux(<Home/>);
    const homeElement = getByTestId('transactionListData');
    //console.log(homeElement);
    expect(homeElement).toBeInTheDocument();
})

test('should render list of transactions', () => {
    const {getByTestId} = renderwithRedux(<Home/>);
    const flowerElement = getByTestId('transaction0');
    const salaryElement = getByTestId('transaction1');
    const bookElement = getByTestId('transaction2');
    const cameraElement = getByTestId('transaction3');
    expect(flowerElement).toBeInTheDocument();
    expect(salaryElement).toBeInTheDocument();
    expect(bookElement).toBeInTheDocument();
    expect(cameraElement).toBeInTheDocument();
})

test('should render list of transactions1', () => {
    const {getByTestId} = renderwithRedux(<Home/>);
    const flowerElement = getByTestId('btntransaction0');
    const salaryElement = getByTestId('transaction1');
    const bookElement = getByTestId('transaction2');
    const cameraElement = getByTestId('transaction3');
    expect(flowerElement).toBeInTheDocument();
    expect(salaryElement).toBeInTheDocument();
    expect(bookElement).toBeInTheDocument();
    expect(cameraElement).toBeInTheDocument();
})

test('delete fire event', () => {
    const {getByTestId, container } = renderwithRedux(<Home/>);
    const deleteflowerElement = getByTestId('btntransaction0');
    console.log(deleteflowerElement)
    fireEvent.click(deleteflowerElement);
    container.querySelector('.minus')
    expect(container.firstChild).toMatchSnapshot()
})

test('should render Add Transacion Component', () => {
    const {getByTestId} = renderwithRedux(<Home/>);
    const addTransactionElement = getByTestId('addTransaction');
    expect(addTransactionElement).toBeInTheDocument();
})

test("render AddTransaction" , () => {
    const handleSubmitMock = jest.fn();
    const transaction = { };
    const { getByTestId } = renderwithRedux(<AddTransaction transaction= {transaction} clearData={handleSubmitMock}/>);
    const formElement = getByTestId('btnAddTransaction');

    expect(formElement).toBeInTheDocument();
    //fireEvent.click(formElement);
    //expect(handleSubmitMock).toBeCalled();
})

test("Click on AddTransaction button" , () => {
    const handleSubmitMock = jest.fn();
    const transaction = { };
    const { getByTestId } = renderwithRedux(<AddTransaction transaction= {transaction} clearData={handleSubmitMock}/>);
    const formElement = getByTestId('btnAddTransaction');
    fireEvent.click(formElement);
    expect(handleSubmitMock).toBeCalled();
})