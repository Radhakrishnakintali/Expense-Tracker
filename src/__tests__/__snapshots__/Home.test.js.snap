// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dispatches the correct action and payload1 1`] = `
Array [
  Object {
    "payload": 1,
    "type": "DELETE_TRANSACTION",
  },
]
`;

exports[`delete fire event 1`] = `
<div
  style="width: 100%;"
>
  <div>
    <div
      data-testid="transactionListData"
    >
      <div
        width="100%"
      >
        <div
          class="font-weight-normal mb-2"
          style="padding-top: 20px;"
        >
          financial_flow_chart
        </div>
        <svg
          height="600px"
          style="margin: auto; display: block;"
          width="600px"
        >
          <g
            style="mix-blend-mode: multiply;"
          >
            <rect
              fill="#8dd3c7"
              height="583.9999999999998"
              width="15"
              x="292.5"
              y="3.794193362993041"
            >
              <title>
                Budget
              </title>
            </rect>
            <rect
              fill="#a9a1b3"
              height="424.72727272727263"
              width="15"
              x="1"
              y="1"
            >
              <title>
                Salary
              </title>
            </rect>
            <rect
              fill="#f0d1e1"
              height="21.23636363636365"
              width="15"
              x="584"
              y="3.858254561310341"
            >
              <title>
                Book
              </title>
            </rect>
            <rect
              fill="#ffed6f"
              height="159.27272727272714"
              width="15"
              x="1"
              y="435.72727272727263"
            >
              <title>
                Camera
              </title>
            </rect>
            <path
              d="M16,213.36363636363637C154.25,213.36363636363637,154.25,216.15782972662942,292.5,216.15782972662942"
              fill="lightgrey"
              style="fill: none; stroke-opacity: 0.5; stroke: #a9a1b3; stroke-width: 424.72727272727275;"
            >
              <title>
                2000
              </title>
            </path>
            <path
              d="M307.5,14.412375181174859C445.75,14.412375181174859,445.75,14.47643637949216,584,14.47643637949216"
              fill="lightgrey"
              style="fill: none; stroke-opacity: 0.5; stroke: #8dd3c7; stroke-width: 21.236363636363638;"
            >
              <title>
                100
              </title>
            </path>
            <path
              d="M16,515.3636363636363C154.25,515.3636363636363,154.25,508.1578297266294,292.5,508.1578297266294"
              fill="lightgrey"
              style="fill: none; stroke-opacity: 0.5; stroke: #ffed6f; stroke-width: 159.27272727272728;"
            >
              <title>
                750
              </title>
            </path>
          </g>
        </svg>
      </div>
      <h3>
        history
      </h3>
      <ul
        class="list"
        id="list"
      >
        <li
          class="plus"
          data-testid="transaction0"
        >
          Salary
           
          <span>
            +
            $
            2000
          </span>
          <button
            class="delete-btn"
            data-testid="btntransaction0"
          >
            x
          </button>
          <button
            type="button"
          >
            Edit
          </button>
        </li>
        <li
          class="minus"
          data-testid="transaction1"
        >
          Book
           
          <span>
            -
            $
            100
          </span>
          <button
            class="delete-btn"
            data-testid="btntransaction1"
          >
            x
          </button>
          <button
            type="button"
          >
            Edit
          </button>
        </li>
        <li
          class="plus"
          data-testid="transaction2"
        >
          Camera
           
          <span>
            +
            $
            750
          </span>
          <button
            class="delete-btn"
            data-testid="btntransaction2"
          >
            x
          </button>
          <button
            type="button"
          >
            Edit
          </button>
        </li>
      </ul>
    </div>
    <div
      data-testid="addTransaction"
      style="width: 40%;"
    >
      <h3>
        add_new_transaction
      </h3>
      <form
        id="form"
      >
        <div
          class="form-control"
        >
          <label
            for="text"
          >
            text
          </label>
          <input
            id="text"
            placeholder="enter_text..."
            type="text"
            value=""
          />
        </div>
        <div
          class="form-control"
        >
          <label
            for="amount"
          >
            amount
             
            <br />
            (
            negative
             - 
            expense
            , 
            positive
             - 
            income
            )
          </label>
          <input
            id="amount"
            placeholder="enter_amount..."
            type="number"
            value="0"
          />
        </div>
        <button
          class="btn"
          data-testid="btnAddTransaction"
          style="background-color: lightblue;"
        >
          add_transaction
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`fire event 1`] = `
<div>
  <div>
     Cash Flow Chart
  </div>
  <div
    data-testid="transactionListData"
  >
    <h3>
      History
    </h3>
    <ul
      class="list"
      id="list"
    >
      <li
        class="plus"
        data-testid="transaction0"
      >
        Salary
         
        <span>
          +
          $
          300
        </span>
        <button
          class="delete-btn"
          data-testid="btntransaction0"
        >
          x
        </button>
        <button
          type="button"
        >
          Edit
        </button>
      </li>
      <li
        class="minus"
        data-testid="transaction1"
      >
        Book
         
        <span>
          -
          $
          10
        </span>
        <button
          class="delete-btn"
          data-testid="btntransaction1"
        >
          x
        </button>
        <button
          type="button"
        >
          Edit
        </button>
      </li>
      <li
        class="plus"
        data-testid="transaction2"
      >
        Camera
         
        <span>
          +
          $
          150
        </span>
        <button
          class="delete-btn"
          data-testid="btntransaction2"
        >
          x
        </button>
        <button
          type="button"
        >
          Edit
        </button>
      </li>
    </ul>
  </div>
  <div
    data-testid="addTransaction"
  >
    <h3>
      Add new transaction
    </h3>
    <form
      id="form"
    >
      <div
        class="form-control"
      >
        <label
          for="text"
        >
          Text
        </label>
        <input
          id="text"
          placeholder="Enter text..."
          type="text"
          value=""
        />
      </div>
      <div
        class="form-control"
      >
        <label
          for="amount"
        >
          Amount 
          <br />
          (negative - expense, positive - income)
        </label>
        <input
          id="amount"
          placeholder="Enter amount..."
          type="number"
          value="0"
        />
      </div>
      <button
        class="btn"
        data-testid="btnAddTransaction"
      >
        Add transaction
      </button>
    </form>
  </div>
</div>
`;
